ingester/ingester.go

```diff
@@ -487,42 +489,70 @@ func (i *Ingester) flushAllUsers(immediate bool) {
 	i.userStateLock.Unlock()
 
 	for id, state := range userState {
-		i.flushUser(id, state, immediate)
-	}
-}
-
-func (i *Ingester) flushUser(userID string, userState *userState, immediate bool) {
-	for pair := range userState.fpToSeries.iter() {
-		i.flushSeries(userState, pair.fp, pair.series, immediate)
+		for pair := range state.fpToSeries.iter() {
+			state.fpLocker.Lock(pair.fp)
+			i.sweepSeries(id, pair.fp, pair.series, immediate)
+			state.fpLocker.Unlock(pair.fp)
+		}
 	}
 
-	// TODO: this is probably slow, and could be done in a better way.
 	i.userStateLock.Lock()
-	if userState.fpToSeries.length() == 0 {
-		delete(i.userState, userID)
+	for id, state := range userState {
+		if state.fpToSeries.length() == 0 {
+			delete(i.userState, id)
+		}
 	}
 	i.userStateLock.Unlock()
 }
 
-func (i *Ingester) flushSeries(u *userState, fp model.Fingerprint, series *memorySeries, immediate bool) {
-	// Enqueue this series flushing if the oldest chunk is older than the threshold
-
-	u.fpLocker.Lock(fp)
+// sweepSeries schedules a series for flushing based on a set of criterea
+//
+// NB we don't close the head chunk here, as the series could wait in the queue
+// for some time, and we want to encourage chunks to be as full as possible.
+func (i *Ingester) sweepSeries(userID string, fp model.Fingerprint, series *memorySeries, immediate bool) {
 	if len(series.chunkDescs) <= 0 {
-		u.fpLocker.Unlock(fp)
 		return
 	}
 
 	lastTime := series.lastTime
-	flush := immediate || len(series.chunkDescs) > 1 || model.Now().Sub(lastTime) > i.cfg.MaxChunkIdle
-	u.fpLocker.Unlock(fp)
+	flush := i.shouldFlushSeries(series, immediate)
 
 	if flush {
 		flushQueueIndex := int(uint64(fp) % uint64(i.cfg.ConcurrentFlushes))
-		i.flushQueues[flushQueueIndex].Enqueue(&flushOp{lastTime, u.userID, fp, immediate})
+		i.flushQueues[flushQueueIndex].Enqueue(&flushOp{lastTime, userID, fp, immediate})
 	}
 }
 
+func (i *Ingester) shouldFlushSeries(series *memorySeries, immediate bool) bool {
+	// Series should be scheduled for flushing if they have more than one chunk
+	if immediate || len(series.chunkDescs) > 1 {
+		return true
+	}
+
+	// Or if any chunks need flushing
```

Comment:
Maybe "Or if the only existing chunk needs flushing"? If that's actually the intent, then maybe rejig the code below to be

```golang
if len(series.chunkDescs) == 0 {
  return false
}
return i.shouldFlushChunk(series.chunkDescs[0])
```